Support .only for all files

naptowncode opened the pull request:
My suggested solution for #549. Needs unit tests, of course.

naptowncode commented:
@sindresorhus @vdemedes @jamestalmage Added unit test. Should be ready to go.

BarryThePenguin commented:
Is there a test for multiple exclusive tests in multiple files?

naptowncode commented:
The test uses three files: One file that has only an exclusive test, one file that has an exclusive test and a non-exclusive test, one file that has two non-exclusive test. Total of five tests across three files with two exclusive tests in different files.

novemberborn commented:
I'm wondering how this should work with `--watch` mode. Let's say we have two test files: `foo.js` and `bar.js`. The latter contains a `.only`.

Initially all test files are run and because `bar.js` contains a `.only`, only that test is actually executed.

Subsequently `foo.js` is changed. The watcher will run just `foo.js` which does not contain a `.only`. Now all tests in `foo.js` are executed.

Later a non-test file is changed and all test files are rerun. Only the `.only` test from `bar.js` is executed.

This isn't necessarily bad but it may be surprising to some. I wonder if `.only` should be sticky so that when changes are detected the watcher still only executes `.only` tests. But then if you remove the `.only` it starts executing all tests again. Similarly adding a `.only` switches back into the sticky mode.

This is a little tricky to implement though #544 requires similar per-test-file housekeeping.

Anyway not meant as a blocker to this PR landing. We can discuss in a separate issue if people think `.only` should stick across watcher runs.

novemberborn commented on api.js:
```diff
@@ -108,6 +109,15 @@ Api.prototype._handleExceptions = function (data) {
 };
 
 Api.prototype._handleStats = function (stats) {
+	if (this.hasExclusive && !stats.hasExclusive) {
```
Could you add a comment as to why the function is returning here? It's because if we're running exclusive tests, but the test file does not include exclusive tests, then we don't want those tests to be counted, right?

naptowncode replied:
Correct, although I'm not super keen on just dropping this info on the floor. I just went for consistency with how non-exclusive tests are dropped within `test-collection`. I'll add a comment to that effect.

That said, I would favor changing the reporting completely as suggested by #471.

novemberborn commented on api.js:
```diff
@@ -108,6 +109,15 @@ Api.prototype._handleExceptions = function (data) {
+
+	if (!this.hasExclusive && stats.hasExclusive) {
+		this.hasExclusive = true;
+		this.testCount = 0;
```
Similarly why is this being reset? Because we've already counted tests we're now not going to run, right? I appreciate `test-collection` has similar logic without comments but I reckon it'd be easier to understand for the next person if we'd explain it here.

naptowncode replied:
Yup, same logic. Again, not too keen on it but couldn't think of a more elegant solution. Suggestions welcome!

novemberborn replied:
Nah it's good enough as is, just requires some extra thinking.

novemberborn commented on index.js:
```diff
@@ -78,12 +79,18 @@ globals.setImmediate(function () {
-	process.on('ava-run', function () {
+	process.on('ava-run', function (options) {
+		if (options.runOnlyExclusive && !hasExclusive) {
+			exit();
```
Maybe pass `runOnlyExclusive` to `runner.run()`? It can still create its `stats` object inside the `run()` method, but bail out if it has no exclusive tests.

naptowncode replied:
Good call. That's almost certainly a better approach.

novemberborn commented on test/api.js:
```diff
@@ -616,6 +616,26 @@ test('test file with only skipped tests does not create a failure', function (t)
+		path.join(__dirname, 'fixture/one-pass-one-fail.js')
+	];
+
+	var api = new Api(files);
```
`master` moves the `files` argument to `api.run(files)`. You'll have to rebase this branch, sorry :)

novemberborn commented:
Solution makes sense to me. Great job given the amount of (necessary) indirection involved in coordinating test runs :+1:

naptowncode commented:
I'm not actually sure how it works with `--watch`. At the very least I should add some documentation for that.

novemberborn commented:
> I'm not actually sure how it works with --watch. At the very least I should add some documentation for that.

Watch itself isn't documented yet so don't worry about it. We can get this in and then discuss repercussions in a new issue.

naptowncode commented:
@novemberborn I've incorporated those tweaks. If everything looks good, merging this would make it easier to work on #471. Thanks!

sindresorhus commented:
LGTM. @vdemedes @jamestalmage ?

vdemedes commented:
LGTM from me! @naptowncode could you please resolve the merge conflict? Big thanks for your work, this is an important change!

novemberborn commented on lib/runner.js:
```diff
@@ -88,14 +90,26 @@ Runner.prototype._addTestResult = function (result) {
+
+Runner.prototype.run = function (options) {
+	options = options || {};
+	this._reset();
```
You no longer need the `_reset()` method now that `run` is always called.

novemberborn commented on lib/fork.js:
```diff
@@ -134,14 +134,16 @@ module.exports = function (file, opts) {
 
-	promise.run = function () {
+	promise.run = function (options) {
+		options = options || {};
```
Given that this is an internal method we should assume it's always called with an `options` object. You'll have to update `test/fork.js` though. Maybe some other places too.

novemberborn replied:
Yes. Tests aside there's only one place the "run chain" is called and it's from the Api.

novemberborn commented on api.js:
```diff
@@ -177,24 +187,30 @@ Api.prototype.run = function (files) {
 							return;
 						}
 
+						runOnlyExclusive = runOnlyExclusive || stats.hasExclusive;
```
I like this approach! Seems cleaner than the `this.hasExclusive` approach on top, though it can't fix the `testCount`. Maybe the test count should be incremented here as well and we can do away with `_handleStats`?

Note though that `tryRun` is also called when on `test.catch()`. You'd have to wrap it there to avoid passing the error as the `stats` object and then add a guard for `stats` being undefined.

Or maybe stick with the `_handleStats()` approach. Just not both :wink:

naptowncode replied:
It does seem redundant but it's to avoid responsibility creep. `_handleStats()` is concerned with gathering the total `testCount` before tests run while this method is waiting for each test to emit stats before starting them all. Probably the `stats` event itself is overloaded by using it for both purposes.

I'll leave `_handleStats()` as it is but wrap the `tryRun()` on `test.catch()`.

novemberborn replied:
Sure, but then why not use `self.hasExclusive` here?

naptowncode replied:
To keep to test execution concerns contained in `tryRun()` and separate from the bookkeeping concerns in `_handleStats()`. But it's not a big hangup for me. I'll go ahead and update `tryRun()` to use `self.hasExclusive`.

novemberborn commented on index.js:
```diff
@@ -78,13 +79,18 @@ globals.setImmediate(function () {
+			runOnlyExclusive: eventOptions.runOnlyExclusive
+		};
+
+		runner.run(runOptions).then(exit);
```
Just use `eventOptions` here I think. And rename it to `options`.

naptowncode commented:
@novemberborn @vdemedes @sindresorhus Updated and rebased. Also added a unit test that should have been there earlier for the `.runOnlyExclusive` option to `Runner.run()`.

sindresorhus commented:
@naptowncode Tests are failing on Travis it seems: https://travis-ci.org/sindresorhus/ava/builds/112949149

novemberborn commented:
@sindresorhus that's just one test in Node 0.10 though.

novemberborn commented:
@naptowncode LGTM!

vdemedes commented:
Thank you so much, @naptowncode!

novemberborn commented:
Please see #593 for how `.only` should interact in watch mode.
